📝 Typst template documentation (#1179)
jupyter-book/mystmd: packages/jtex/docs/create-a-typst-template.md

---
title: Create a Typst Template

> description: jtex templates have a template.yml, template.typ, and any other images, class or definition files required for the template to render.

Reviewer: ```suggestion
description: Typst templates have a template.yml, template.typ, and any other images, class or definition files required for the template to render.
```

thumbnail: ./thumbnails/create-a-typst-template.png
---

A `jtex` template contains everything necessary to create a $\LaTeX$ document, including a `template.yml`, the main `template.typ`, and any associated files such as classes (`*.cls`), definitions (`*.def`), or images (`*.png`).

> These $\LaTeX$ templates are data-driven, in that they record all of the options in a `template.yml` which you create as you are working through moving your $\LaTeX$ document to a `jtex` template.

Reviewer: ```suggestion
These Typst templates are data-driven, in that they record all of the options in a `template.yml` which you create as you are working through moving your Typst document to a `jtex` template.
```

ultimately this means wrapping your `parts` in conditional statements, such as `[# if parts.abstract #]` or providing fallbacks.
```

## Update the authors and affiliations

The authors and affiliations are usually the hardest part to template as many journals do these differently.
If you are looking for inspiration, take a look at some of the existing templates in the [myst-templates](https://github.com/myst-templates) organization on GitHub.

For example, to create the following author/affiliations list:

```latex
\authors{First Author\affil{1}, Author\affil{1,2}}

\affiliation{1}{First Affiliation}
\affiliation{2}{Second Affiliation}
```

The `jtex` template is as follows:

```latex
\authors{
[#- for author in doc.authors -#]
[-- author.name --]
[#- if author.affiliations -#]
\affil{
[-- author.affiliations|join(",", "index") --]
}
[#- endif -#]
[#- if not loop.last #], [# endif -#]
[#- endfor -#]
}

[# for affiliation in doc.affiliations #]
\affiliation{[-affiliation.index-]}{[-affiliation.value-]}
[# endfor #]

> ```

Reviewer: TODO: refer to the frontmatter plugin from Curvenote for typst

To automatically find packages, ensure that your `files` list is up to date (including any style or other classes), and `jtex` will (naively) parse these and provide warnings. To put the automatically found packages into your `template.yml`, **save** and run:

```bash
jtex check --fix
```

This will overwrite your `template.yml` with all packages found and there should be very few issues found automatically by `jtex check`.

```{figure} ./images/jtex-check-fix.png
:width: 100%

Using `jtex check --fix` will fix as many errors as possible with your `template.yml`!
```

## Build with content

Your template should now be in a place where it can be used to render content. For this we will use the `myst` command line tool.

Create a markdown file with your content, with some frontmatter that looks like:

```yaml
---
title: My article title
exports:
  - format: tex
    template: ../my_template # The folder with your template.yml in it
    draft: true # Any options
    journal_name: Nature
    keypoint: I know how to make a MyST Markdown template.
authors:
  - name: Rowan Cockett
    orcid: 0000-0002-7859-8394
    affiliations:
      - Curvenote
keywords:
  - MyST Templates
---
```

> Ensure that the exports list has a `format: tex` in it. To also have your `parts` defined, use blocks with JSON metadata:

Reviewer: ```suggestion
Ensure that the exports list has a `format: typst` in it. To also have your `parts` defined, use blocks with JSON metadata:
```

```markdown
+++ {"part": "abstract"}

This is your abstract!

+++

# Introduction

Other content!
```

You can now render your document with:

```bash
myst build my-document.md --tex
```

> By default these are put in a `_build` folder. If you want to control that, use the `output:` field in the appropriate export. If you have $\LaTeX$ installed, you can also try changing the format to `pdf` or `pdf+tex` to keep the source files. See [](/guide/creating-pdf-documents) for more information on using MyST templates for $\LaTeX$.

Reviewer: ```suggestion
By default these are put in a `_build` folder. If you want to control that, use the `output:` field in the appropriate export. If you have Typst installed, you can also try changing the format to `pdf` or `pdf+typst` to keep the source files. See [](/guide/creating-pdf-documents) for more information on using MyST templates for Typst.
```
